src/snowflake/cli/api/utils/rendering.py

```diff
@@ -88,24 +97,152 @@ def jinja_render_from_file(
         print(rendered_result)
 
 
-class _AttrGetter:
-    def __init__(self, data_dict):
-        self._data_dict = data_dict
+def _add_project_context(
+    external_data: Dict, project_definition: ProjectDefinition
+) -> Dict:
+    """
+    Updates the external data with variables from snowflake.yml definition file.
+    """
+    if _CONTEXT_KEY in external_data:
+        raise ClickException(
+            f"{_CONTEXT_KEY} in user defined data. The `{_CONTEXT_KEY}` variable is reserved for CLI usage."
+        )
+
+    context_data = _resolve_variables_in_project(project_definition)
+
+    context_data.update(external_data)
+    return context_data
+
+
+def _remove_ctx_env_prefix(text: str) -> str:
+    prefix = "ctx.env."
+    if text.startswith(prefix):
+        return text[len(prefix) :]
+    return text
+
+
+def string_includes_template(text: str) -> bool:
```

Comment:
There is an issue loading yaml using the &{} template, so we will hit the chicken and egg issue here. We need to parse the yaml before we expand the templates.
Should we change the syntax - choose something parsable by yaml for project file and eventually deprecate the sql-only syntax?
Cam suggested ${}. Let me know what you think

Reply:
@sfc-gh-turbaszek This seems to be a critical problem with the chosen syntax. Not argument against remaining compatible with the legacy tool in SQL text, but we need to ensure that we don't end up breaking yaml parsing/generating tools because it's the _only_ syntax we support. We already know that customers like to generate a number of their project files, so that seems like an important requirement. Thoughts?

Reply:
I think I agree with @sfc-gh-turbaszek that we can support using variables only as part of strings specified in quotes and cast strings properly in models if needed.

But we have to confirm that it works correctly with multiline strings which do not require quotes.

Reply:
We have to think long term here IMHO. We should be allowing templating in files other than the project definition file (it's on NADE's roadmap for the near future anyway). What if someone is using a numerical port number in a container spec? We can't assume strings are the only use case.

Also, we should not assume a specific parser and/or generator is used. What we make our users write should be valid YAML so that they are free to load their files in any editor, IDE or tool they choose, and it won't fail on them.

Reply:
So the solution you propose is to use different templating for snowflake.yml. I'm fine with that.  But let's be mindful we will end up with 4 templates:
- {{ ... }} for server side
- &{ ... }  for SQL
- ${ ... }  for snowflake.yaml
- ??? for creating templates of yamls that can be rendered, this syntaxt has to be different than all previous

Reply:
To add on that, the idea of templating is that you replace the template with the value before reading it. In that case, if we don't use the quote, the template marker should be replaced by a non-quoted string, which is a valid YAML - which is why it doesn't seem right to force users with quoting it.

Would we consider the idea of using {{ }} instead, and have snow CLI only parse stuff that starts with ctx. ?

Reply:
Looking for ctx requires parsing before expanding. Plus one can have SQL template in snsowlfake.yml that uses `{{ ctx }}`. Templates within templates within templates 🤯

Reply:
I think we have to parse before expanding, and do it iteratively - at least for snowflake.yaml, because vars are defined there. Could you please elaborate on the second part about sql templates in snowflake.yml?

Reply:
Imagine we will add support for embedding SQL in snowflake.yml. Then users may be tempted to 
```
sql:
   run: "EXECUTE IMMEDIATE  SELECT * FROM {{ ctx }}"
env:
   ...
```
In this way usage of `{{ }}` will be ambiguous. Thus, we need different template pattern.

Reply:
It's the same context, so it shouldn't matter where it's replaced. Or am I missing something?

Reply:
No no, I'm not proposing to use 4 different template syntax options. `{{ }}` is the server side one, so that ship has sailed. I'm proposing that we pick something for the client side that works for both SQL and YAML, and make that the client-side template. Then we also support `&{ }` as a legacy feature (maybe it's even a feature flag?). We monitor its use for deprecation purposes. If you use &{}, you use it exclusively, and only in a SQL file. Everybody else should be using the new and improved syntax that doesn't break YAML parsers. So really, it's only 2 options, other than legacy, which is something we must support regardless.

Reply:
Having legacy for sql, still leaves us with 3 templates (legacy, jinja, something for yaml) + one more if someone wants to generate snowflake.yml, or do I miss something?

Reply:
Why would snowflake.yml be any different? But yes, 3 total, 2 actively supported and 1 legacy. I do think that's the minimal number. Happy to involve the product folks here if you think it would be worthwhile.